packages/compile-solidity/compileWithPragmaAnalysis.js

```diff
@@ -0,0 +1,139 @@
+const CompilerSupplier = require("./compilerSupplier");
+const Config = require("@truffle/config");
+const semver = require("semver");
+const Profiler = require("./profiler");
+const fse = require("fs-extra");
+const { run } = require("./run");
+const OS = require("os");
+const cloneDeep = require("lodash.clonedeep");
+
+const getSemverExpression = source => {
+  return source.match(/pragma solidity(.*);/)[1].trim();
+};
+
+const getSemverExpressions = sources => {
+  return sources.map(source => getSemverExpression(source));
+};
+
+const validateSemverExpressions = semverExpressions => {
+  const { validRange } = semver;
+  for (const expression of semverExpressions) {
+    if (semver.validRange(expression) === null) {
+      const message = `Invalid semver expression (${expression}) found in` +
+        `one of your contract's imports.`;
+      throw new Error(message);
+    }
+  }
+};
+
+// takes an array of versions and an array of semver expressions
+// returns a version of the compiler or undefined if none can be found
+const findNewestSatisfyingVersion = ({ solcReleases, semverExpressions }) => {
+  // releases are ordered from newest to oldest
+  return solcReleases.find(version => {
+    return semverExpressions.every(expression =>
+      semver.satisfies(version, expression)
+    );
+  });
+};
+
+const throwCompilerVersionNotFound = ({ path, semverExpressions }) => {
+  const message =
+    `Could not find a single version of the Solidity compiler that ` +
+    `satisfies the following semver expressions obtained from your source ` +
+    `files' pragma statements: ${semverExpressions.join(" - ")}. ` +
+    `${OS.EOL}Please check the pragma statements for ${path} and its imports.`;
+  throw new Error(message);
+};
+
+const compileWithPragmaAnalysis = async ({ paths, options }) => {
+  const supplierOptions = {
+    events: options.events,
+    solcConfig: options.compilers.solc
+  };
+  const compilerSupplier = new CompilerSupplier(supplierOptions);
+  const { releases } = await compilerSupplier.getReleases();
+
+  // collect sources by the version of the Solidity compiler that they require
+  const versionsAndSources = {};
+  for (const path of paths) {
+    const source = (await options.resolver.resolve(path)).body;
+
+    const parserVersion = findNewestSatisfyingVersion({
```

Comment:
Hm, this means that we won't find a suitable compiler for parsing unless there's a single version they all have in common, right?  So that won't allow mixing of e.g. `pragma solidity ^0.5.0` and `pragma solidity ^0.6.0`? :-/  That's troublesome...

I guess in order for it to be otherwise... I guess the parser itself would have to inspect the pragma and load an appropriate Solidity compiler according to it.  Yeah I suppose that is a bit troublesome...

Reply:
Note that this happens for each "path". This means that every source file will be analyzed this way to find an appropriate version of the parser.

Reply:
Ah, I see, thanks!

But, hm, there's still a problem here, isn't there?  Because it's trying to determine what parser to use for a whole tree based on a single file -- the newest one that works with that file had better also work with all its dependencies.

OTOH, maybe that isn't a problem -- if the newest one that works with that file doesn't work with its dependencies, it should have specified its pragma better!  Also, we're just talking about getting a version for *parsing* here, and that will likely be fairly robust; parser changes happen but, like, relevant ones shouldn't happen that often, right?

So I guess this is probably fine, at least for an initial version.  Whew, I was worried there!

Reply:
Yeah so this is also "experimental" which makes me feel better about the cracks people might find in this functionality haha.